src/web-extensions.c

```diff
@@ -0,0 +1,719 @@
+                            "sender", call->sender,
+                            "app-id", xdp_app_info_get_id (call->app_info),
+                            "token", session_token,
+                            "connection", connection,
+                            NULL);
+
+  if (session)
+    g_debug ("webextensions session owned by '%s' created", session->sender);
+
+  return (WebExtensionsSession *)session;
+}
+
+static gboolean
+handle_create_session (XdpDbusWebExtensions *object,
+                       GDBusMethodInvocation *invocation,
+                       GVariant *arg_options)
+{
+  Call *call = call_from_invocation (invocation);
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  g_autoptr(GError) error = NULL;
+  Session *session;
+
+  session = (Session *)web_extensions_session_new (arg_options, call, connection, &error);
+  if (!session)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+  if (!session_export (session, &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      session_close (session, FALSE);
+      return TRUE;
+    }
+  session_register (session);
+
+  xdp_dbus_web_extensions_complete_create_session (object, invocation, session->id);
+
+  return TRUE;
+}
+
+static void
+on_server_exited (GPid pid,
+                  gint status,
+                  gpointer user_data)
+{
+  Session *session = user_data;
+  WebExtensionsSession *web_extensions_session = (WebExtensionsSession *)session;
+
+  SESSION_AUTOLOCK (session);
+  web_extensions_session->child_pid = -1;
+  web_extensions_session->child_watch_id = 0;
+  session_close (session, TRUE);
+}
+
+static gboolean
+array_contains (JsonArray *array,
+                const char *value)
+{
+  guint length, i;
+
+  if (array == NULL)
+    return FALSE;
+
+  length = json_array_get_length (array);
+  for (i = 0; i < length; i++)
+    {
+      const char *element = json_array_get_string_element (array, i);
+      if (g_strcmp0 (element, value) == 0)
+        return TRUE;
+    }
+  return FALSE;
+}
+
+static gboolean
+is_valid_name (const char *name)
+{
+  /* This regexp comes from the Mozilla documentation on valid native
+     messaging server names:
+
+     https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_manifests#native_messaging_manifests
+
+     That is, one or more dot-separated groups composed of
+     alphanumeric characters and underscores.
+  */
+  return g_regex_match_simple ("^\\w+(\\.\\w+)*$", name, 0, 0);
+}
+
+static GStrv
+get_manifest_search_path (void)
+{
+  const char *hosts_path_str;
+  g_autoptr(GPtrArray) search_path = NULL;
+
+  hosts_path_str = g_getenv ("XDG_DESKTOP_PORTAL_WEB_EXTENSIONS_PATH");
+  if (hosts_path_str != NULL)
+    return g_strsplit (hosts_path_str, ":", -1);
+
+  /* By default, use the native messaging search paths of Firefox,
```

Comment:
Sorry for neglecting this. My main concern here is that it's not going to stop at just Firefox and Chromium: taking Keepass for example, it has hard coded seven possible locations just for per-user configuration:

https://github.com/keepassxreboot/keepassxc/blob/develop/src/browser/NativeMessageInstaller.cpp#L67-L74

My hope was to have an implementation that would not require changes every time a new browser was adapted to use the portal. In particular, because updates to xdg-desktop-portal on people's systems are far less frequent than updates to applications distributed via Flathub or the Snap Store. I don't think we want to bypass this problem by having the browser pass a search path to the portal, since that would open up a whole new set of problems (e.g. probing .

I'm not sure how realistic the "one browser tricking another browser" case is. In the case of snaps, access to dot files in the user's home directory is forbidden. And if an application has access to `~/.config`, then they can break out of their sandbox by writing a file to `~/.config/autostart` and waiting for the next login.

Lastly, the browser is always going to have to put a level of trust in the portal: it's likely both the manifests and executables are not visible from the browser's namespace, so there is no way it can verify what the portal provides it with. I don't think it makes sense for the browser to be second guessing the portal.